util: fix inspecting of proxy objects

jasnell opened the pull request:
##### Checklist

- [x] tests and code linting passes
- [x] a test and/or benchmark is included
- [x] the commit message follows commit guidelines

##### Affected core subsystem(s)

util

##### Description of change

In certain conditions (see https://github.com/nodejs/node/issues/6464), inspecting a Proxy object can lead to a max call stack error. Avoid that by detecting the Proxy object and outputting information about the Proxy object itself.

Also adds util.isProxy()

Fixes: https://github.com/nodejs/node/issues/6464

/cc @bnoordhuis

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/2428/

TheAlphaNerd commented:
CI is green LGTM

vkurchatkin commented:
-1, the whole point of proxies is being transparent

addaleax commented:
> the whole point of proxies is being transparent

That’s true for the programmatic use of proxies, but `util.inspect` comes with the intention of showing humans what is going on.

jasnell commented:
@addaleax +1 . Note that `getters/setters` are also generally intended to be transparent for most cases but util.inspect will produce output like `{a: [Getter]}` which avoids similar stack-depth side effects... e.g. 

```js
var m = {};
Object.defineProperty(m, 'a', {
  enumerable: true,
  get: function() {
    console.log(this);
    return 5;
  }
});
util.inspect(m);
  // Returns `{a: [Getter]}`
```

jasnell commented:
(@TheAlphaNerd ... what are you doing up at 2am reviewing code I wrote at 1am? Sleep my friend! Sleep!)

cjihrig commented on lib/util.js:
```diff
@@ -785,6 +792,10 @@ exports.isPrimitive = isPrimitive;
+function isProxy(p) {
+  return binding.isProxy(p);
+}
+exports.isProxy = isProxy;
```
If you export this, it becomes public API, and should be documented.

jasnell replied:
+1 ... Forgot to add the doc for it. Will add soon!
On Apr 29, 2016 6:29 AM, "Colin Ihrig" <notifications@github.com> wrote:

> In lib/util.js
> <https://github.com/nodejs/node/pull/6465#discussion_r61577439>:
>
> > @@ -785,6 +792,10 @@ exports.isPrimitive = isPrimitive;
> >
> >  exports.isBuffer = Buffer.isBuffer;
> >
> > +function isProxy(p) {
> > +  return binding.isProxy(p);
> > +}
> > +exports.isProxy = isProxy;
>
> If you export this, it becomes public API, and should be documented.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/6465/files/3d4ff4311a3cf1ff756413756ffc859a663aba99#r61577439>
>

cjihrig commented on test/parallel/test-util-inspect-proxy.js:
```diff
@@ -0,0 +1,22 @@
+assert.deepStrictEqual(target, details.target);
+assert.deepStrictEqual(handler, details.handler);
+
+assert.equal(util.inspect(a),
```
`strictEqual()` please.

cjihrig commented on test/parallel/test-util-inspect-proxy.js:
```diff
@@ -0,0 +1,22 @@
+const handler = {
+  get: function() { console.log(arguments); return 5; }
+};
+const a = new Proxy(target, handler);
```
Can you give this a more meaningful name.

cjihrig commented on test/parallel/test-util-inspect-proxy.js:
```diff
@@ -0,0 +1,22 @@
+
+const target = {};
+const handler = {
+  get: function() { console.log(arguments); return 5; }
```
Can you drop the `console.log()`.

jasnell replied:
Should be able to replace it with `util.inspect(arguments[0])` but the
point of the test is that this bit of code is *not* run. I included it here
because it was part of the original bug report. Throwing in here would work
too.
On Apr 29, 2016 6:31 AM, "Colin Ihrig" <notifications@github.com> wrote:

> In test/parallel/test-util-inspect-proxy.js
> <https://github.com/nodejs/node/pull/6465#discussion_r61577734>:
>
> > @@ -0,0 +1,22 @@
> > +'use strict';
> > +
> > +require('../common');
> > +const assert = require('assert');
> > +const util = require('util');
> > +
> > +const target = {};
> > +const handler = {
> > +  get: function() { console.log(arguments); return 5; }
>
> Can you drop the console.log().
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/6465/files/3d4ff4311a3cf1ff756413756ffc859a663aba99#r61577734>
>

jasnell commented:
@cjihrig ... updated! PTAL

jasnell commented:
New CI after update: https://ci.nodejs.org/job/node-test-pull-request/2432/

cjihrig commented on doc/api/util.md:
````diff
@@ -478,6 +478,23 @@ util.isPrimitive(new Date())
+
+```js
+const util = require('util');
+
````
Unnecessary blank line.

cjihrig commented on test/parallel/test-util-inspect-proxy.js:
```diff
@@ -0,0 +1,22 @@
+};
+const proxyObj = new Proxy(target, handler);
+
+assert(util.isProxy(proxyObj));
```
Just to be safe, this should probably `assert.strictEqual()` with `true`.

cjihrig replied:
A similar assertion for a `false` return should be added.

cjihrig commented on test/parallel/test-util-inspect-proxy.js:
```diff
@@ -0,0 +1,22 @@
+// Inspecting the proxy should not actually walk it's properties
+assert.doesNotThrow(() => util.inspect(proxyObj));
+
+const details = process.binding('util').getProxyDetails(proxyObj);
```
Similarly, it would be nice to verify the return value when this is called without a proxy object.

jasnell commented:
@cjihrig ... done! :-)

@vkurchatkin ... +1 on it being semver-major (unfortunately).

jasnell commented:
@vkurchatkin ... would this be a bit more palatable if there were an option on `util.inspect` to turn *off* this special treatment of Proxy objects? e.g. `util.inspect(proxyObj, {unwrapProxy: false});`

cjihrig commented on test/parallel/test-util-inspect-proxy.js:
```diff
@@ -0,0 +1,29 @@
+require('../common');
+const assert = require('assert');
+const util = require('util');
+const process_util = process.binding('util');
```
This should probably be camelCase.

cjihrig commented:
LGTM pending CI, but it should sit for a couple days.

vkurchatkin commented:
> would this be a bit more palatable if there were an option on util.inspect to turn off

@jasnell after some thinking, I agree with your reasoning. Having an option is nice, but in practice no one uses them

jasnell commented:
@cjihrig +1 for letting it sit a couple days. There's definitely no rush on it.
@vkurchatkin ... awesome.

jasnell commented:
Squashed, new CI: https://ci.nodejs.org/job/node-test-pull-request/2435/

jasnell commented:
CI is green. Will land on monday if there are no objections. /cc @nodejs/ctc

Fishrock123 commented:
Looks like this conflates two things:

1. fixing inspect (which shouldn't be major)
2. adding `util.isProxy()` (which we **should not do** given we don't even know what the hell do do with the existing functions: https://github.com/nodejs/node/pull/6235)

Do not land in this state.

jasnell commented:
So that's a -1 for existing isProxy(), no worries there, that's a minor bit
that can be easily dropped.

Any constructive comments or objections to the actual fix?

It can very well be considered a major given that it changes the output
entirely for a certain class of objects. A case can be made, tho, that we
had never intended to support proxy here so it can be a minor. Either way
works for me because either way the bug gets fixed.
On Apr 29, 2016 4:31 PM, "Jeremiah Senkpiel" <notifications@github.com>
wrote:

Looks like this conflates two things:

   1. fixing inspect (which shouldn't be major)
   2. adding util.isProxy() (which we *should not do* given we don't even
   know what the hell do do with the existing functions: #6235
   <https://github.com/nodejs/node/pull/6235>)

Do not land in this state.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/nodejs/node/pull/6465#issuecomment-215907288>

jasnell commented:
@vkurchatkin ... since you're the one who marked this major, are you ok with downgrading this to semver-minor or even patch?

jasnell commented:
Updated! New CI: https://ci.nodejs.org/job/node-test-pull-request/2441/

jasnell commented:
@TheAlphaNerd @cjihrig @vkurchatkin @bnoordhuis @mscdex  ... was looking this over and the performance hit was just too much to be doing the isProxy check by default. Added a `showProxy` switch that is off by default, so in order to get this new behavior you'd have to pass `{showProxy: true}` to `util.inspect()`. Added caching also to avoid duplicate lookups.

bnoordhuis commented on lib/util.js:
```diff
@@ -122,6 +122,7 @@ function inspect(obj, opts) {
   // default options
   var ctx = {
     seen: [],
+    proxyCache: new WeakMap(),
```
Seems a little wasteful to instantiate a new WeakMap that's not used unless showProxy is true.  I'd just set it to undefined here and instantiate it lazily inside formatValue().

jasnell replied:
+1

bnoordhuis commented on lib/util.js:
```diff
@@ -241,6 +243,23 @@ function inspectPromise(p) {
 
 function formatValue(ctx, value, recurseTimes) {
+
+  if (ctx.showProxy && typeof value === 'object') {
```
E.g.:
```js
var proxyCache = ctx.proxyCache;
if (proxyCache === undefined)
  proxyCache = ctx.proxyCache = new WeakMap();
```

bnoordhuis commented on lib/util.js:
```diff
@@ -241,6 +243,23 @@ function inspectPromise(p) {
+
+  if (ctx.showProxy && typeof value === 'object') {
+    const inProxyCache = ctx.proxyCache.has(value);
+    var proxy = inProxyCache ? ctx.proxyCache.get(value) : undefined;
```
It is necessary to .has() before .get()?  The latter returns undefined when the entry doesn't exist.

jasnell replied:
See https://github.com/nodejs/node/pull/6465#discussion_r61670211

bnoordhuis commented on lib/util.js:
```diff
@@ -241,6 +243,23 @@ function inspectPromise(p) {
+      return 'Proxy ' + formatValue(ctx, proxy, recurseTimes);
+    } else if (!inProxyCache) {
+      if (binding.isProxy(value)) {
+        proxy = binding.getProxyDetails(value);
```
You can save a binding call by writing it like this:
```js
proxy = binding.getProxyDetails(value);
if (proxy !== undefined) {
  // ...
}
```
```

bnoordhuis commented on lib/util.js:
```diff
@@ -241,6 +243,23 @@ function inspectPromise(p) {
+        ctx.proxyCache.set(value, proxy);
+        return 'Proxy ' + formatValue(ctx, proxy, recurseTimes);
+      } else {
+        ctx.proxyCache.set(value, undefined);
```
This is pointless, isn't it?

jasnell replied:
Unfortunately, no. This is here to record the object's that we know *aren't* proxies in case we encounter them again while we are recursing through. This is why I do the `has` check above before the `get` also. Consider:

```js
var a = new Proxy({}, {});
var b = { };
var c = [ b, a, b, a ];
```

When I pass this to `util.inspect()`, util.inspect does not have any idea which objects are Proxies and which are normal objects until it does the isProxy call, which is expensive. If formatValue has already checked `b` above and has determined that it's not a Proxy, if we don't record that knowledge somehow we end up having to do the check again when `b` comes back around. This is largely because `util.inspect()` is not very efficient at caching stuff that's already seen. There's some memoization happening later on that I can see about refactoring, but without recording that we've already ruled an object out, this ends up being much slower.

bnoordhuis commented on src/node_util.cc:
```diff
@@ -37,6 +39,21 @@ using v8::Value;
+
+  Local<Object> ret = Object::New(env->isolate());
+  ret->Set(env->handler_string(), proxy->GetHandler());
+  ret->Set(env->target_string(), proxy->GetTarget());
```
`Boolean::New(env->isolate(), proxy->IsRevoked())` might be of interest as well, although I guess you can also infer that by checking if the handler is non-null.

Aside, it might be a little cheaper to return an array instead of a dictionary object.

mscdex replied:
Is it still (even) cheaper to pass an array/object from js?

jasnell replied:
Ok, can do.

bnoordhuis commented on test/parallel/test-util-inspect-proxy.js:
```diff
@@ -0,0 +1,34 @@
+assert.deepStrictEqual(handler, details.handler);
+
+assert.strictEqual(util.inspect(proxyObj, {showProxy: true}),
+    'Proxy { handler: { get: [Function] }, target: {} }');
```
Can you line up the arguments?

Fishrock123 commented:
> It can very well be considered a major given that it changes the output
entirely for a certain class of objects. A case can be made, tho, that we
had never intended to support proxy here so it can be a minor. Either way
works for me because either way the bug gets fixed.

I don't actually disagree but I think it is more important to fix it in current, also we've previously treated these even as patch fixes. https://github.com/nodejs/node/pulls?page=1&q=is%3Apr+is%3Aclosed+util+inspect&utf8=%E2%9C%93

Fishrock123 commented on doc/api/util.md:
```diff
@@ -179,6 +179,9 @@ formatted string:
    defined on the objects being inspected won't be called. Defaults to `true`.
 
+ - `showProxy` - if `true`, then objects that are Proxy objects will be
+   introspected to show their `target` and `hander` objects.
```
defaults to...?

jasnell commented:
@bnoordhuis ... updated! PTAL ... unfortunately it's still dog slow even with your suggested improvements.

Fishrock123 commented:
Does this mean it defaults to `false` in `console.log()`? Shouldn't it be on?

jasnell commented:
Yes, it defaults to `false` in console.log and *should* be `false`. The reason is that the check is *expensive*. We end up have to check *every* object, array and function as it passes through because there's no other way to determine if it's a Proxy object or not. Calling down to `process.binding('util').getProxyDetails()` *every* time we `console.log()` an object is not something I want to do.

mscdex commented:
I agree it should be `false` by default as long as it's as expensive to check.

Fishrock123 commented:
Oh shoot. I agree defaulting to false makes sense for that.

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/2444/

bnoordhuis commented on lib/util.js:
```diff
@@ -241,6 +242,41 @@ function inspectPromise(p) {
 function formatValue(ctx, value, recurseTimes) {
+
+  if (ctx.showProxy &&
+      (typeof value === 'object' || typeof value === 'function')) {
```
You should check that `typeof value === 'object' && value !== null`, I think?

Style nit: can you drop the blank line?

jasnell replied:
Ack

bnoordhuis commented on lib/util.js:
```diff
@@ -241,6 +242,41 @@ function inspectPromise(p) {
+    var proxy = undefined;
+    var proxyCache = ctx.proxyCache;
+    if (!proxyCache)
+      proxyCache = ctx.proxyCache = new WeakMap();
```
Can this be a `Map` instead of a `WeakMap`?

mscdex replied:
Or an empty object?

jasnell replied:
Can, yes. Will change.

bnoordhuis commented on lib/util.js:
```diff
@@ -241,6 +242,41 @@ function inspectPromise(p) {
+      if (proxy) {
+        // We know for a fact that this isn't a Proxy.
+        // Mark it as having already been evaluated.
+        proxyCache.set(proxy, undefined);
```
Why are you storing the array that .getProxyDetails() returns as a key?  I would expect this to be `proxyCache.set(value, proxy)`.  The comment seems wrong as well.  This whole `if` statement seems like it could be dropped, actually.

jasnell replied:
Look at lines 273 and 276.

bnoordhuis commented on src/node_util.cc:
```diff
@@ -37,6 +39,21 @@ using v8::Value;
+
+  Local<Proxy> proxy = args[0].As<Proxy>();
+
+  Local<Array> ret = Array::New(env->isolate(), 2);
```
If you use `args.GetIsolate()` here, you can drop the call to `Environment::GetCurrent()` a few lines up.

jasnell replied:
Ack

bnoordhuis commented on test/parallel/test-util-inspect-proxy.js:
```diff
@@ -0,0 +1,84 @@
+// This is here to test that the internals are working correctly.
+const details = processUtil.getProxyDetails(proxyObj);
+assert.deepStrictEqual(target, details[0]);
+assert.deepStrictEqual(handler, details[1]);
```
Shouldn't these be `assert.strictEqual()`?  I assume you want to test for identity equality, not value equality.

jasnell replied:
Ack

jasnell commented:
@bnoordhuis ... updated

bnoordhuis commented on lib/util.js:
```diff
@@ -241,6 +242,46 @@ function inspectPromise(p) {
+        // for it to get proper formatting, and because
+        // the target and handle objects also might be
+        // proxies... it's unfortunate but necessary.
+        proxyCache.set(proxy, undefined);
```
I understand the logic but I don't think the 'necessary' is accurate; if you skipped adding `proxy` to the cache here, the algorithm would still work, only less efficiently.  The next iteration would check the cache, miss, call .getProxyDetails() for the details array, and add the `undefined` return value to the cache a few lines below.

jasnell replied:
Necessary in the sense of avoiding an extraneous call to getProxyDetails... Which I would consider to be a bug since the point is to only check when we *don't* know for sure. ;)

bnoordhuis commented:
LGTM

Fishrock123 commented:
CI after nits.. seems fine to me if others are happy with the details: https://ci.nodejs.org/job/node-test-pull-request/2459/

jasnell commented:
@TheAlphaNerd and @cjihrig ... still LGTY?

TheAlphaNerd commented:
LGTM

cjihrig commented:
Yes, LGTM

jasnell commented:
Landed in ba6196f.
Thank you all.
